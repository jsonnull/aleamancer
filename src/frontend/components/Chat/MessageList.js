// @flow
import * as React from 'react'
import styled from 'styled-components'
import type { Message } from 'common/types'
import MessageView from './Message'

const MessagesWrapper = styled.div`
  min-height: 0;
  flex: 1;
`
const Messages = styled.div`
  overflow-y: ${props => (props.isPinned ? 'scroll' : 'hidden')};
  overflow-x: hidden;
  width: 315px;
  height: 100%;

  display: flex;
  flex-direction: column;
  flex: ${props => (props.isPinned ? '1' : 'none')};

  align-items: ${props => (props.isPinned ? '' : 'flex-end')};
`

type Props = {
  messages: Array<Message>,
  isPinned: boolean
}

export default class MessageList extends React.Component<Props> {
  timer: IntervalID
  scroll: ?HTMLElement

  componentDidMount() {
    // Every minute, update chat timestamps
    this.timer = setInterval(() => this.forceUpdate(), 60000)
  }

  componentDidUpdate() {
    if (this.scroll) {
      this.scroll.scrollTop += 10000
    }
  }

  componentWillUnmount() {
    clearInterval(this.timer)
  }

  render() {
    const { isPinned } = this.props
    return (
      <MessagesWrapper>
        <Messages
          isPinned={isPinned}
          innerRef={el => {
            this.scroll = el
          }}
        >
          {this.props.messages.map(message => (
            <MessageView
              key={message.id}
              message={message}
              isPinned={isPinned}
            />
          ))}
        </Messages>
      </MessagesWrapper>
    )
  }
}
