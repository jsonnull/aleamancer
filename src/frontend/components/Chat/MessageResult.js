// @flow
import React from 'react'
import styled from 'styled-components'
import type { MessageResult, Roll } from 'common/types'

const RollSpan = styled.span`
  font-weight: bold;
`

type RollProps = {
  result: number,
  operation: '+' | '-',
  color?: 'inherit' | 'green' | 'red',
  showOperation?: boolean
}
const Result = (props: RollProps) => {
  const { result, operation, color = 'inherit', showOperation = true } = props

  const shownOperation = showOperation ? ` ${operation} ` : ''

  return (
    <span>
      {shownOperation}
      <RollSpan style={{ color }}>{result}</RollSpan>
    </span>
  )
}

type RollContainerProps = {
  roll: Roll | number,
  isFirst: boolean
}
const RollContainer = (props: RollContainerProps) => {
  const { roll, isFirst } = props

  // Handle the general case where the roll is a constant
  if (typeof roll === 'number') {
    return <Result result={roll} operation="+" showOperation={!isFirst} />
  }

  const { die, result, operation } = roll

  let color = 'inherit'
  if (result === die) {
    color = 'green'
  } else if (result === 1) {
    color = 'red'
  }

  return (
    <Result
      result={result}
      color={color}
      operation={operation}
      showOperation={!isFirst}
    />
  )
}

const ResultsContainer = styled.div`
  padding: 0 1rem 0.2rem;
`
const Total = styled.span`
  font-weight: bold;
`

type Props = {
  result: ?MessageResult
}
const Results = (props: Props) => {
  if (!props.result) {
    return null
  }

  const results = props.result
  const total = results.reduce((acc, roll) => {
    if (typeof roll === 'number') {
      return acc + roll
    } else {
      return acc + roll.result
    }
  }, 0)

  return (
    <ResultsContainer>
      <i className="fa fa-cube" />{' '}
      {results.map((result, i) => (
        <span key={i}>
          <RollContainer roll={result} isFirst={i === 0} />
        </span>
      ))}
      {' = '}
      <Total>{total}</Total>
    </ResultsContainer>
  )
}

export default Results
