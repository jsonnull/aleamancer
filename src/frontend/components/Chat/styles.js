// @flow
import styled from 'styled-components'

export const Container = styled.div`
  background-color: ${props => props.theme.background};
  width: 320px;
  overflow: hidden;
  border-radius: ${props => (props.isPinned ? '0' : '5px')};
  box-shadow: ${props =>
    props.isPinned ? '' : '0 3px 5px 0 rgba(0, 0, 0, 0.2)'};

  height: ${props => (props.isPinned ? '100%' : 'auto')};
  position: ${props => (props.isPinned ? 'relative' : 'absolute')};
  top: ${props => (props.isPinned ? '0' : 'auto')};
  right: ${props => (props.isPinned ? '0' : '1rem')};
  bottom: ${props => (props.isPinned ? '0' : '1rem')};

  display: flex;
  flex-direction: column;
`

export const Loading = styled.div`
  padding: 10px;
`

export const Error = Loading
