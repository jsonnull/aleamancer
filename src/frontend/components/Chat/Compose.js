// @flow
import React from 'react'
import styled from 'styled-components'
import { fontSize, fonts } from 'frontend/styles/common'

const Form = styled.form``

const Input = styled.textarea.attrs({
  type: 'text'
})`
  border: 0;
  height: 50px;
  line-height: 3rem;
  padding: 1rem;
  display: block;
  width: 100%;
  resize: none;
  color: ${props => props.theme.color};
  background-color: ${props => props.theme.background};
  border-radius: 5px;
  font-size: ${fontSize.normal};
  font-family: ${fonts.body};
`

type Props = {
  onSend: Function,
  isPinned: boolean
}

type State = {
  value: string,
  height: number
}

const defaultHeight = 50

export default class Compose extends React.Component<Props, State> {
  focused: boolean
  messageQueue: Array<string>
  autogrow: ?HTMLElement

  constructor(props: Props) {
    super(props)

    this.state = {
      value: '',
      height: 0
    }
    this.focused = false
    this.messageQueue = []
  }

  handleChange(event: Object) {
    const value = event.target.value

    const setHeightImmediate = value.length > this.state.value.length

    this.setState({
      value
    })

    if (setHeightImmediate) {
      this.calculateHeight()
    } else {
      this.setState({ height: 0 })
      window.requestAnimationFrame(() => this.calculateHeight())
    }
  }

  calculateHeight() {
    if (!this.autogrow) {
      return
    }

    const autogrow = this.autogrow
    const scrollHeight = autogrow.scrollHeight

    let height = 0
    if (scrollHeight !== defaultHeight) {
      height = scrollHeight
    }

    this.setState({
      height
    })
  }

  handleKeyUp(event: Object) {
    if (event.key == 'Enter') {
      this.handleSubmit()
    }
  }

  handleSubmit() {
    this.messageQueue.push(this.state.value)
    this.props.onSend(this.state.value)
    this.setState({
      value: '',
      height: 0
    })
  }

  render() {
    const { isPinned } = this.props

    const style = {}
    if (this.state.height !== 0) {
      Object.assign(style, { height: this.state.height + 'px' })
    }

    return (
      <Form>
        <Input
          placeholder="Send a message..."
          value={this.state.value}
          innerRef={el => {
            this.autogrow = el
          }}
          style={style}
          isPinned={isPinned}
          onChange={e => this.handleChange(e)}
          onKeyUp={e => this.handleKeyUp(e)}
        />
      </Form>
    )
  }
}
