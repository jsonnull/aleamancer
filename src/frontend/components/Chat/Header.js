// @flow
import React from 'react'
import styled from 'styled-components'
import Tooltip from '../Tooltip'
import { fontSize } from 'frontend/styles/common'

const Container = styled.div`
  background-color: ${props =>
    props.isPinned ? props.theme.background : props.theme.backgroundInverted};
  display: flex;
  flex-direction: row;
`

const Toggle = styled.div`
  display: block;
  font-size: ${fontSize.small};
  color: ${props =>
    props.isPinned ? props.theme.color : props.theme.colorInverted};
  margin-right: 1rem;
  padding-left: 1rem;
  cursor: pointer;

  i {
    transform: ${props => (props.isPinned ? 'none' : 'rotate(45deg)')};
  }
`

const Outer = styled.div`
  margin-left: auto;
`

type Props = {
  isPinned: boolean,
  setChatPinned: Function
}

const Header = (props: Props) => {
  const { isPinned, setChatPinned } = props

  let toggleChat = isPinned ? (
    <i className="fa fa-chevron-left" />
  ) : (
    <i className="fa fa-thumb-tack" />
  )

  const tooltip = isPinned ? 'Unpin Chat' : 'Pin Chat'

  return (
    <Container isPinned={isPinned}>
      <Outer>
        <Tooltip placement="left" content={tooltip}>
          <Toggle onClick={() => setChatPinned(!isPinned)} isPinned={isPinned}>
            {toggleChat}
          </Toggle>
        </Tooltip>
      </Outer>
    </Container>
  )
}

export default Header
