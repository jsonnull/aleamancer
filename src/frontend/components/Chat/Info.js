// @flow
import React from 'react'
import styled from 'styled-components'
import { fontSize } from 'frontend/styles/common'

const Container = styled.div`
  background-color: ${props => props.theme.backgroundSecondary};
  line-height: 2.5rem;
  color: ${props => props.theme.colorMedium};
  font-size: ${fontSize.small};
  padding: 0 1rem;
`

const Info = () => (
  <Container>
    {'/roll 1d6, '}
    <strong>*bold*</strong> <em>_italic_</em>
  </Container>
)

export default Info
