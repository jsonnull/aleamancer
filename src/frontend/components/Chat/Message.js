// @flow
import React from 'react'
import styled from 'styled-components'
import Timeago from 'timeago.js'
import type { Message } from 'common/types'
import { fontSize } from 'frontend/styles/common'
import MessageResult from './MessageResult'

type Props = {
  message: Message,
  isPinned: boolean
}

const MessageContainer = styled.div`
  padding: 0.5rem 0;
  border-bottom: 1px solid ${props => props.theme.borderColor};
  display: inline-block;
  width: 100%;

  &:first-child {
    margin-top: auto;
  }
`

const From = styled.div`
  font-weight: bold;
  font-size: ${fontSize.small};
  color: ${props => props.theme.colorSecondary};
  padding: 0.2rem 1rem 0;
`

const Sent = styled.span`
  color: #999;
  font-weight: normal;
  font-style: italic;
`

const Text = styled.div`
  padding: 0 1rem 0.2rem;
  color: ${props => props.theme.color};
`

export default class MessageView extends React.Component<Props> {
  render() {
    const { isPinned } = this.props
    const { from, timestamp, text, result } = this.props.message
    const hoverText = new Date(timestamp)
    const formattedDate = new Timeago().format(timestamp)
    return (
      <MessageContainer isPinned={isPinned}>
        <From>
          {from} <Sent title={hoverText}>{formattedDate}</Sent>
        </From>
        <Text>{text}</Text>
        <MessageResult result={result} />
      </MessageContainer>
    )
  }
}
