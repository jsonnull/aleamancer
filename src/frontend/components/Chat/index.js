// @flow
import React from 'react'
import styled from 'styled-components'
import type { Message } from 'common/types'
import Header from './Header'
import Info from './Info'
import Compose from './Compose'
import MessageList from './MessageList'

const CHAT_WIDTH = '320px'

const Container = styled.div`
  background-color: ${props => props.theme.background};
  width: ${props => props.chatWidth};
  overflow: hidden;
  border-radius: ${props => (props.isPinned ? '0' : '5px')};
  box-shadow: ${props =>
    props.isPinned ? '' : '0 3px 5px 0 rgba(0, 0, 0, 0.2)'};

  height: ${props => (props.isPinned ? '100%' : 'auto')};
  position: ${props => (props.isPinned ? 'relative' : 'absolute')};
  top: ${props => (props.isPinned ? '0' : 'auto')};
  right: ${props => (props.isPinned ? '0' : '1rem')};
  bottom: ${props => (props.isPinned ? '0' : '1rem')};

  display: flex;
  flex-direction: column;
`

type Props = {
  messages: Array<Message>,
  isPinned: boolean,
  setChatPinned: Function,
  sendMessage: Function
}

class Chat extends React.Component<Props> {
  messageQueue: []

  sendMessage = (text: string) => {
    this.props.sendMessage(text.trim())
  }

  render() {
    const { isPinned, messages, setChatPinned } = this.props

    const shownMessages = isPinned ? messages : messages.slice(-4)

    return (
      <Container chatWidth={CHAT_WIDTH} isPinned={isPinned}>
        <Header isPinned={isPinned} setChatPinned={setChatPinned} />
        <MessageList messages={shownMessages} isPinned={isPinned} />
        <Compose onSend={this.sendMessage} isPinned={isPinned} />
        <Info />
      </Container>
    )
  }
}

export default Chat
