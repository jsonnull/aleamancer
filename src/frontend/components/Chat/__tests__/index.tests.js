// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import Chat from '../index.js'

describe('Chat component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Chat
          messages={[]}
          isPinned={false}
          setChatPinned={() => {}}
          sendMessage={() => {}}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
