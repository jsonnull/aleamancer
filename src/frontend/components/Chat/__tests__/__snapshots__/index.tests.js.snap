// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat component renders correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c3 {
  display: block;
  font-size: 1.3rem;
  margin-right: 1rem;
  padding-left: 1rem;
  cursor: pointer;
}

.c3 i {
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c2 {
  margin-left: auto;
}

.c7 {
  line-height: 2.5rem;
  font-size: 1.3rem;
  padding: 0 1rem;
}

.c6 {
  border: 0;
  height: 50px;
  line-height: 3rem;
  padding: 1rem;
  display: block;
  width: 100%;
  resize: none;
  border-radius: 5px;
  font-size: 1.5rem;
  font-family: -apple-system,BlinkMacSystemFont,  "Segoe UI","Roboto","Oxygen","Ubuntu","Cantarell","Fira Sans","Droid Sans",  "Helvetica Neue",sans-serif;
}

.c4 {
  min-height: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 {
  overflow-y: hidden;
  overflow-x: hidden;
  width: 315px;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c0 {
  width: 320px;
  overflow: hidden;
  border-radius: 5px;
  box-shadow: 0 3px 5px 0 rgba(0,0,0,0.2);
  height: auto;
  position: absolute;
  top: auto;
  right: 1rem;
  bottom: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <div>
          <div
            className="c3"
            onClick={[Function]}
          >
            <i
              className="fa fa-thumb-tack"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c4"
  >
    <div
      className="c5"
    />
  </div>
  <form
    className=""
  >
    <textarea
      className="c6"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="Send a message..."
      style={Object {}}
      type="text"
      value=""
    />
  </form>
  <div
    className="c7"
  >
    /roll 1d6, 
    <strong>
      *bold*
    </strong>
     
    <em>
      _italic_
    </em>
  </div>
</div>
`;
